### PR TITLE
Melee Rework: Stage One

### PR DESCRIPTION
**Pull request by SugarBone**
## About The Pull Request
**!!!NEEDS TESTING!!!**

Adds Swing Based AOE attacks, adds broad one-target attacks. Enables double-tact weaponry. Stumble now works only if legs were hit. Attacks for dislocation now switched to Grab intent. Changed the variables on weaponry to enable such attacks (this has to be adjusted - the current state is not necessarily the correct one, more of a placeholder). In order to do a Swing attack, you need to wield an appropriate weapon and used it on harm intent - it hits the target and two mobs/structures adjacent to it. In order to use a broad attack, you need to use an appropriate weapon on Disarm intent - it attacks three bodyparts (two of which are randomly selected), with damage divided by 3 for each. Some weapons have a forced broad attack. Double tact addresses the issue of unbalanced weapons, requiring the user to click the target twice within the span of 10 seconds. 

## Why It's Good For The Game

Spices up the melee combat, makes it more viable and interesting. Completes intended design.

## Changelog
:cl:
add: Added swing based attacks, broad attacks, and double tact weapons
tweak: Adjust appropriate weapons to their attack types, changes stumble to only work when legs are targeted; attacks for dislocation switched to grab intent
/:cl:

### CODE REVIEW

File: code/modules/mob/living/carbon/human/human_defense.dm [hunk 1]
```diff
@@ -63,10 +63,10 @@ meteor_act
 	var/hit_dirs = (r_dir in cardinal) ? r_dir : list(r_dir & NORTH|SOUTH, r_dir & EAST|WEST)
 
 	var/stumbled = FALSE
-
-	if(prob(60 - stats.getStat(STAT_TGH)))
-		stumbled = TRUE
-		step(src, pick(cardinal - hit_dirs))
+	if(hit_zone == BP_L_LEG || hit_zone == BP_R_LEG)
```
Comment (Humonitarian) [inline code/modules/mob/living/carbon/human/human_defense.dm:66]: There should be a define for a list of the two legs combined. BP_LEGS I recall?

**Comment by SugarBone** [comment]
Added the modifier var to all procs, should be safe for testing now

File: code/_onclick/item_attack.dm [hunk 2]
```diff
@@ -36,10 +36,77 @@ avoid code duplication. This includes items that may sometimes act as a standard
...
+					if(holdinghand == slot_l_hand)
+						src.tileattack(user, L, modifier = 1)
+						src.tileattack(user, C, modifier = 0.8)
+						src.tileattack(user, R, modifier = 0.6)
+						return
+					else if(holdinghand == slot_r_hand)
+						src.tileattack(user, R, modifier = 1)
+						src.tileattack(user, C, modifier = 0.8)
+						src.tileattack(user, L, modifier = 0.6)
+						return
```
Comment (SirRichardFrancis) [inline code/_onclick/item_attack.dm:105]:
```suggestion
	if(doubletact)
		if(ready))
			ready = 0
		else
			user.visible_message(SPAN_DANGER("[user] raises \his [src]"))
			ready = 1
			var/endtime = world.time + 100
			while(world.time < endtime)
				sleep(1)
				if(!ready)
					return
				if(!is_equipped())
					ready = 0
					return
			user.visible_message(SPAN_DANGER("[user] lowers \his [src]"))
			ready = 0
			return

	if(ishuman(user) && can_swing && wielded) // Swinging
		var/mob/living/carbon/human/H = user
		if(H.a_intent == I_HURT)
			var/x_difference = A.x - H.x
			var/y_difference = 
			var/z_level = A.z
			var/holdinghand = user.get_inventory_slot(src)
			var/turf/R
			var/turf/C
			var/turf/L
			C = locate(A.x, A.y, z_level)
			switch(x_difference)
				if(0)
					R = locate((A.x + 1), A.y, z_level)
					L = locate((A.x - 1), A.y, z_level)
				if(1)
					switch(y_difference)
						if(0)
							R = locate(A.x, (A.y - 1), z_level)
							L = locate(A.x, (A.y + 1), z_level)
						if(1)
							R = locate(A.x, (A.y - 1), z_level)
							L = locate((A.x - 1), A.y, z_level)
						if(-1)
							R = locate((A.x - 1), A.y, z_level)
							L = locate(A.x, (A.y + 1), z_level)
				if(-1)
					switch(y_difference)
						if(0)
							R = locate(A.x, (A.y + 1), z_level)
							L = locate(A.x, (A.y - 1), z_level)
						if(1)
							R = locate((A.x + 1), A.y, z_level)
							L = locate(A.x, (A.y - 1), z_level)
						if(-1)
							R = locate(A.x, (A.y + 1), z_level)
							L = locate((A.x + 1), A.y, z_level)
			if(holdinghand == slot_l_hand)
				tileattack(user, L, modifier = 1)
				tileattack(user, C, modifier = 0.8)
				tileattack(user, R, modifier = 0.6)
				return
			else if(holdinghand == slot_r_hand)
				tileattack(user, R, modifier = 1)
				tileattack(user, C, modifier = 0.8)
				tileattack(user, L, modifier = 0.6)
				return
```
I don't like this code, but at least here is unfucked syntax.

File: code/_onclick/item_attack.dm [hunk 3]
```diff
@@ -85,13 +152,24 @@ avoid code duplication. This includes items that may sometimes act as a standard
 		return TRUE
 	return I.attack(src, user, user.targeted_organ)
 
+//Handles AOE attacking on tiles
+/obj/item/proc/tileattack(mob/living/user, turf/targetarea, var/modifier = 1)
+	if(istype(targetarea, /turf/simulated/wall))
+		var/turf/simulated/W = targetarea
+		return W.attackby(src, user)
+	for(var/obj/structure/S in targetarea)
+		if (S.density && !istype(S, /obj/structure/table))
```
Comment (SirRichardFrancis) [inline code/_onclick/item_attack.dm:161]:
```suggestion
		if(S.density && !istype(S, /obj/structure/table))
```
Read `CONTRIBUTING.dm`.

File: code/_onclick/item_attack.dm [hunk 4]
```diff
@@ -85,13 +152,24 @@ avoid code duplication. This includes items that may sometimes act as a standard
...
+			S.attackby(src, user)
+	for(var/mob/living/M in targetarea)
+		src.attack(M, user, user.targeted_organ, modifier)
```
Comment (SirRichardFrancis) [inline code/_onclick/item_attack.dm:164]:
```suggestion
		attack(M, user, user.targeted_organ, modifier)
```

File: code/_onclick/item_attack.dm (hunk 3 above)
Comment (SugarBone) [reply code/_onclick/item_attack.dm:174]: No

**Comment by SugarBone** [comment]
I approve of the /var/modifier changes, but I am not going to bother fixing it all again, since it has no effect anyways

**Comment by SirRichardFrancis** [comment]
>image

Okie dokie, I'll spoon-feed to you relevant parts then.

```
Name your proc parameters properly to prevent name conflicts.
If in doubt, use the prefix _ to clearly indicate an input parameter.
Do not use src.var if it can be helped.
```
image

All of those is redundant and serve no purpose whatsoever.

```
Spaces are needed between function agruments (declaration and definition).
Spaces are needed between the binary operator and arguments.
Space is not needed when the operation is unary.
Spaces are not needed near brackets.
Spaces are needed around assignment operator.
```

image

```
Don't have unnecessary return calls or return meaningless data.
If there's nothing after a return, and its not returning a specific value, you don't need a return at all.
```
image
image

```
del() usage is deprecated, use qdel() instead.
```

image

```
Boolean variables and return values should use TRUE and FALSE constans instead of 1 and 0.
```
image

**Comment by SirRichardFrancis** [comment]
Also would look a lot cleaner if instead of `if(!(src.ready)` you write `if(ready)` and flip cases. Capiche?

**Comment by SugarBone** [comment]
Purely cosmetic standards to which the codebase doesn't even adhere

**Comment by SirRichardFrancis** [comment]
> Purely cosmetic standards to which the codebase doesn't even adhere, plus the maintainers can just merge your changes if it is that big of a deal

Nope, your shitcode not only looks bad, but actually affects performance in a negative way, most notable misuse of `del()`.
It is very easy to fix, so what exactly is stopping you?

**Comment by MLGTASTICa** [comment]
> Purely cosmetic standards to which the codebase doesn't even adhere, plus the maintainers can just merge your changes if it is that big of a deal

We use qdel instead of del because byond has a hard time deleting shit at the moment you want it to, so don't use del.

**Comment by SugarBone** [comment]
> > Purely cosmetic standards to which the codebase doesn't even adhere, plus the maintainers can just merge your changes if it is that big of a deal
> 
> Nope, your shitcode not only looks bad, but actually affects performance in a negative way, most notable misuse of `del()`. It is very easy to fix, so what exactly is stopping you?

Are you saying that using 1 and 0 instead of TRUE and FALSE and spaces in front of if statements affect performance, yeah, right.

**Comment by SirRichardFrancis** [comment]
> Are you saying that using 1 and 0 instead of TRUE and FALSE and spaces in front of if statements affect performance, yeah, right.

_*sigh_

No, I am saying that using TRUE and FALSE for bools is good for code readability and required by Eris coding policies, as well as other things I've mentioned.
Redundant self-references and returns may not have performance impact to speak of, but again, code readability is important, as well as adhering to style guide of a project you're contributing to.

**Comment by SugarBone** [comment]
> > Are you saying that using 1 and 0 instead of TRUE and FALSE and spaces in front of if statements affect performance, yeah, right.
> 
> _*sigh_
> 
> No, I am saying that using TRUE and FALSE for bools is good for code readability and required by Eris coding policies, as well as other things I've mentioned. Redundant self-references and returns may not have performance impact to speak of, but again, code readability is important, as well as adhering to style guide of a project you're contributing to.

If the codebase even attempted to adhere to those standards, I would have too. It doesn't, hence I don't have to.

**Comment by SirRichardFrancis** [comment]
> If the codebase even attempted to adhere to those standards, I would have too. It doesn't, hence I don't have to.

We are genuinely trying to, but alas, there are people like you and a huge mess of Bay12 legacy code.
All of the necessary changes have been made for you and delivered on a silver plate with justification and explanation for each of them.
If your ego won't let you to accept a code review, then dunno, go touch grass or something.

**Review by Firefox13 — CHANGES_REQUESTED** [review]
I can sign (almost) under every KIROV's word.
Even if the guidelines "aren't adhered to" they are not to be outright ignored and refused.

**Comment by Holvr0** [comment]
Somme comments after today's testing:

**DOUBLE TACT WEAPONS**

Either I don't get the purpose of double-tact weapons or they're awful to use.

The requirement of double clicking essentially halves your DPS in actual combat. It's all cool when the target is stock still and you can doubleclick them, but when both of you move *and* you include the lag in the equation, it becomes severly unfun to sacrifice every other click for raising your weapon instead of hitting. Other weapons hit every time.

In order to "balance" this, the double-tact weapons would have to be super OP in terms of damage, which in turn would make them into funni meme assassin weapons to only use on an unaware target because you can reliably doubleclick them.


**IF** that functionality was available to **all** weapons as a sort of "fire mode", then ok, it'd have some use. But shoehorning certain weapons into it will make them really unfun. Especially compared to the swing weapons, which also brings me to...


**SWING WEAPONS**

PvP functionality is good and fun to use, but swing weapons can't hit roaches. Even on a direct click on roach sprite. You have to let go of your other hand (make it unwielded) to hit roaches.

Swinging also interacts super weird doors - trying to bash doors with a wielded weapon is hitting a wall next to the door. What is weirder, that even applies to weapons that do not have swing in the first place - for example, the IH tactical hammer is double-tact instead of swing, yet trying to bash doors with it makes it try to do a swing and hit a wall instead. Weird interaction.

**Comment by SugarBone** [comment]
> Somme comments after today's testing:
> 
> **DOUBLE TACT WEAPONS**
> 
> Either I don't get the purpose of double-tact weapons or they're awful to use.
> 
> The requirement of double clicking essentially halves your DPS in actual combat. It's all cool when the target is stock still and you can doubleclick them, but when both of you move _and_ you include the lag in the equation, it becomes severly unfun to sacrifice every other click for raising your weapon instead of hitting. Other weapons hit every time.
> 
> In order to "balance" this, the double-tact weapons would have to be super OP in terms of damage, which in turn would make them into funni meme assassin weapons to only use on an unaware target because you can reliably doubleclick them.
> 
> **IF** that functionality was available to **all** weapons as a sort of "fire mode", then ok, it'd have some use. But shoehorning certain weapons into it will make them really unfun. Especially compared to the swing weapons, which also brings me to...
> 
> **SWING WEAPONS**
> 
> PvP functionality is good and fun to use, but swing weapons can't hit roaches. Even on a direct click on roach sprite. You have to let go of your other hand (make it unwielded) to hit roaches.
> 
> Swinging also interacts super weird with doors - trying to bash doors with a wielded weapon is hitting a wall next to the door. What is weirder, that even applies to weapons that do not have swing in the first place - for example, the IH tactical hammer is double-tact instead of swing, yet trying to bash doors with it makes it try to do a swing and hit a wall instead. Weird interaction.

I agree somewhat with the problems of double tact; it could be potentially postponed with review of the designed problems. As for the problems with the swing weapons, attacking doors should be an easy enough fix (an oversight on my part). I would have to review the roach problem.

**Comment by MemedHams** [comment]
as I said before, double-tact has promise, but it has to be converted more into a "readied/heavy" attack system. Give a visual cue above someone when they're swinging their weapon, allow them to hit for boosted power/ap. Currently it feels more like input lag then a "reading attack" effect.

The swing hits are glorious with the new animations, they just need to be fixed to hit nonhuman mobs.

**Comment by SirRichardFrancis** [comment]
nanakocode

**Comment by SugarBone** [comment]
Closing pool request as there are too many issues, caused by some questionable decisions. I will reupload an earlier revision with some necessary up-to-date changes soon.

**Pull request closed without merging**